include/dxc/DXIL/DxilFunctionProps.h

```diff
@@ -25,6 +25,93 @@ class Constant;
 } // namespace llvm
 
 namespace hlsl {
+
+// SM 6.6 allows WaveSize specification for only a single required size.
+// SM 6.8+ allows specification of WaveSize as a min, max and preferred value.
+struct DxilWaveSize {
+  unsigned Min = 0;
+  unsigned Max = 0;
+  unsigned Preferred = 0;
+
+  DxilWaveSize() = default;
+  DxilWaveSize(unsigned min, unsigned max = 0, unsigned preferred = 0)
+      : Min(min), Max(max), Preferred(preferred) {}
+  DxilWaveSize(const DxilWaveSize &other) = default;
+  DxilWaveSize &operator=(const DxilWaveSize &other) = default;
+  bool operator==(const DxilWaveSize &other) const {
+    return Min == other.Min && Max == other.Max && Preferred == other.Preferred;
+  }
+
+  // Encode valid DxilWaveSize, handling potential degenerate cases.
+  static DxilWaveSize Encode(unsigned min, unsigned max = 0,
+                             unsigned preferred = 0) {
+    if (max == min)
+      max = 0;
+    if (max == 0 && preferred == min)
+      preferred = 0;
+    return DxilWaveSize(min, max, preferred);
+  }
+
+  // Valid non-zero values are powers of 2 between 4 and 128, inclusive.
+  static bool IsValidValue(unsigned Value) {
+    return (Value >= 4 && Value <= 128 && ((Value & (Value - 1)) == 0));
+  }
+  // Valid encodings:
+  //  0, 0, 0: Not defined
+  //  Min, 0, 0: single WaveSize (SM 6.6/6.7)
+  //  Min, Max (> Min), 0 or Preferred (>= Min and <= Max): Range (SM 6.8+)
+  enum class ValidationResult {
+    Success,
+    InvalidMin,
+    InvalidMax,
+    InvalidPreferred,
+    MaxOrPreferredWhenUndefined,
+    PreferredWhenNoRange,
+    MaxEqualsMin,
+    MaxLessThanMin,
+    PreferredOutOfRange,
+  };
+  ValidationResult Validate() const {
+    if (Min == 0) { // Not defined
+      if (Max != 0 || Preferred != 0)
+        return ValidationResult::MaxOrPreferredWhenUndefined;
+    } else if (!IsValidValue(Min)) {
+      return ValidationResult::InvalidMin;
+    } else if (Max == 0) { // single WaveSize (SM 6.6/6.7)
```

Comment:
per the spec, max == 0 should be an error regardless

Reply:
No, in the internal canonical representation, Max == 0 indicates a single wave size (no range).
This is just as it was in the prior version, but wrapped in a struct.

Validation or diagnostics for additional constraints on other forms exists elsewhere, where it belongs.

In SemaHLSL the rules for the attribute are checked, leveraging this method for commonalities.
For the cases where you have the attribute and used explicit Min, Max, or Preferred == 0, there is explicit code to translate that to invalid Min/Max/Preferred:
https://github.com/microsoft/DirectXShaderCompiler/pull/6172/files#diff-b5ff0eb7c221e9ad11a754e889ea567c35d83ad9f51e89e77ec844ea4fd8eebbR12638

For the case where you have metadata, we could argue whether it's legal to have metadata that explicitly defines a degenerate case, but if we disallow, it means you couldn't use just one structure of metadata for WaveSize and minimally modify the module in the latest metadata form to represent each valid state.  We don't prevent this kind of thing in other cases - we don't force you to omit an extended attribute definition when the value in the definition is equal to the default.  In any case, that discussion doesn't impact this code, but impacts whether DxilValidation should have extra checks for metadata.

There are no changes from the prior version in this regard.